### PR TITLE
[new feature] Add new features to iterative trimmer

### PR DESCRIPTION
**Pull request by AndrewLevin**
Adds four new arguments to the iterative trimmer that turn on and adjust special features:

1.  `sigmaOffset` 
2. `highNoiseCut`
3. `highTrimCutoff`
4. `highTrimWeight`

## Description
The iterative trimming algorithm now does the following at each iteration:

1. Take an scurve

2. For each channel such that noise is less than `highNoiseCut`, calculate the following quanity: scurve mean plus `sigmaOffset` times scurve sigma

3. Take the weighted average of these quantities, where the weight equals 1 if the trim setting from the previous iteration is less than or equal to `highTrimCutoff`, and equals `highTrimWeight` if the trim setting from the previous iteration is greater than `highTrimCutoff`.

4. Use the formula 15 times difference from weighted average to set the trim settings for the next iteration.

### Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Motivation and Context
This was requested by the QC8 team: https://github.com/cms-gem-daq-project/vfatqc-python-scripts/issues/283

## How Has This Been Tested?
I have used it to trim the coffin with three iterations and found the final scurves look similar to previous trimmed scurves on the coffin:

Summary

## Checklist:

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: iterativeTrim.py [hunk 1]
```diff
@@ -168,6 +168,12 @@ def writeChConfig(chConfig,vfat,vfatID,trimDAC,trimPol,mask,maskReason):
     parser.add_argument("-d","--debug", action="store_true", help="Prints additional debugging information")
     parser.add_argument("-l","--latency", type=int, help="CFG_LATENCY value to be used",default=33)
     parser.add_argument("-m","--maxIter", type=int, help="Maximum number of iterations to perform (e.g. number of scurves to take)", default=4)
+    
+    parser.add_argument("--sigmaOffset", type=int, help="Will align the mean - sigmaOffset*sigma", default=0)
+    parser.add_argument("--highTrimCutoff", type=int, help="Will weight channels that have a trim value above this (when set to 63, has no effect)", default=63)
+    parser.add_argument("--highTrimWeight", type=int, help="Will apply this weight to channels that have a trim value above the cutoff", default=50)    
+    parser.add_argument("--highNoiseCut", type=float, default=1.5, help="Threshold in fC for masking the channel due to high noise")
+
```
Comment (jsturdy) [inline iterativeTrim.py:178]: I have this itching on the back of my neck that makes me wonder whether these types of settings are better stored in some config file that is read by the trim routine, rather than passed around as arguments?

Comment (AndrewLevin) [reply iterativeTrim.py:178]: If it grows to more than 4, I think we can make a config file. I have moved the default values to a separate file, like [anaInfo.py](https://github.com/cms-gem-daq-project/gem-plotting-tools/blob/develop/utils/anaInfo.py) for gem-plotting-tools.

File: iterativeTrim.py [hunk 2]
```diff
@@ -168,13 +168,15 @@ def writeChConfig(chConfig,vfat,vfatID,trimDAC,trimPol,mask,maskReason):
...
     parser.add_argument("-m","--maxIter", type=int, help="Maximum number of iterations to perform (e.g. number of scurves to take)", default=4)
+
+    from gempython.gemplotting.utils.scanInfo  import sigmaOffsetDefault, highTrimCutoffDefault, highTrimWeightDefault, highNoiseCutDefault
     
-    parser.add_argument("--sigmaOffset", type=int, help="Will align the mean - sigmaOffset*sigma", default=0)
-    parser.add_argument("--highTrimCutoff", type=int, help="Will weight channels that have a trim value above this (when set to 63, has no effect)", default=63)
-    parser.add_argument("--highTrimWeight", type=int, help="Will apply this weight to channels that have a trim value above the cutoff", default=50)    
-    parser.add_argument("--highNoiseCut", type=float, default=1.5, help="Threshold in fC for masking the channel due to high noise")
+    parser.add_argument("--sigmaOffset", type=float, help="Will align the mean + sigmaOffset*sigma", default=sigmaOffsetDefault)
+    parser.add_argument("--highTrimCutoff", type=float, help="Will weight channels that have a trim value above this (when set to 63, has no effect)", default=highTrimCutoffDefault)
```
Comment (jsturdy) [inline iterativeTrim.py]: does this really need to be `float`? its representing a discrete quantity, no?

File: iterativeTrim.py [hunk 3]
```diff
@@ -168,13 +168,15 @@ def writeChConfig(chConfig,vfat,vfatID,trimDAC,trimPol,mask,maskReason):
...
+    parser.add_argument("--highTrimWeight", type=float, help="Will apply this weight to channels that have a trim value above the cutoff", default=highTrimWeightDefault)    
+    parser.add_argument("--highNoiseCut", type=float, help="Threshold in fC for masking the channel due to high noise", default=highNoiseCutDefault)
     
-    parser.add_argument("-n","--nevts",type=int,default=100,help="Number of events for each scan position")
+#    parser.add_argument("-n","--nevts",type=int,default=100,help="Number of events for each scan position")
```
Comment (jsturdy) [inline iterativeTrim.py]: why is this removed?

File: run_scans.py [hunk 4]
```diff
@@ -792,10 +792,12 @@ def ultraThreshold(args):
...
 
-    parser_itertrim.add_argument("--sigmaOffset", type=int, help="Will align the mean - sigmaOffset*sigma", default=0)
-    parser_itertrim.add_argument("--highTrimCutoff", type=int, help="Will weight channels that have a trim value above this (when set to 63, has no effect)", default=63)
-    parser_itertrim.add_argument("--highTrimWeight", type=int, help="Will apply this weight to channels that have a trim value above the cutoff", default=50)
-    parser_itertrim.add_argument("--highNoiseCut", type=float, default=1.5, help="Threshold in fC for masking the channel due to high noise")
+    from gempython.gemplotting.utils.scanInfo  import sigmaOffsetDefault, highTrimCutoffDefault, highTrimWeightDefault, highNoiseCutDefault
+    
+    parser_itertrim.add_argument("--sigmaOffset", type=float, help="Will align the mean + sigmaOffset*sigma", default=sigmaOffsetDefault)
+    parser_itertrim.add_argument("--highTrimCutoff", type=float, help="Will weight channels that have a trim value above this (when set to 63, has no effect)", default=highTrimCutoffDefault)
```
Comment (jsturdy) [inline run_scans.py]: re https://github.com/cms-gem-daq-project/vfatqc-python-scripts/pull/286/files/74a11a095887a2276747a81c23e35fc848367c72..18328fc1e51a047bd5baf5fcf1faaa4086aeabaa#r344631291

File: iterativeTrim.py (hunk 2 above)
Comment (AndrewLevin) [reply iterativeTrim.py]: yes, the trim value is an integer, so there is not reason to use a float

File: iterativeTrim.py (hunk 3 above)
Comment (AndrewLevin) [reply iterativeTrim.py]: it should not have been

File: run_scans.py (hunk 4 above)
Comment (AndrewLevin) [reply run_scans.py]: fixed

**Review by jsturdy — DISMISSED** [review]
~~OK for squash~~

**Comment by jsturdy** [comment]
Is this ready for review, or are there further changes coming?
Once you have made a PR that isn't marked WIP, please don't add commits that are not explicitly requested and agreed in the comments

**Comment by AndrewLevin** [comment]
Sorry, I am testing it now on QC8.

**Comment by AndrewLevin** [comment]
This has been tested on a QC8 detector, and after some tweaks, the results look good:

http://cmsonline.cern.ch/cms-elog/1094237

**Review by jsturdy — APPROVED** [review]
ready for squash
